src/arch/x86_64/CodeGen.zig

```diff
@@ -12913,10 +12913,111 @@ fn airShlShrBinOp(self: *CodeGen, inst: Air.Inst.Index) !void {
+                        const tmp_lock = self.register_manager.lockRegAssumeUnused(tmp_reg);
+                        defer self.register_manager.unlockReg(tmp_lock);
+
+                        const lhs_bits: u31 = @intCast(lhs_ty.bitSize(zcu));
+                        const tmp_ty: Type = if (lhs_bits > 64) .usize else lhs_ty;
+                        const off = frame_addr.off + (lhs_bits - 1) / 64 * 8;
+                        try self.genSetReg(
+                            tmp_reg,
+                            tmp_ty,
+                            .{ .load_frame = .{ .index = frame_addr.index, .off = off } },
+                            .{},
+                        );
+                        try self.truncateRegister(lhs_ty, tmp_reg);
+                        try self.genSetMem(
+                            .{ .frame = frame_addr.index },
+                            off,
+                            tmp_ty,
+                            .{ .register = tmp_reg },
+                            .{},
+                        );
+                    },
+                    else => {},
+                }
+                const dst_lock = switch (dst_mcv) {
+                    .register => |reg| self.register_manager.lockRegAssumeUnused(reg),
+                    else => null,
+                };
+                defer if (dst_lock) |lock| self.register_manager.unlockReg(lock);
+
+                // shift right
+                const tmp_mcv = try self.genShiftBinOp(Air.Inst.Tag.shr, null, dst_mcv, rhs_mcv, lhs_ty, rhs_ty);
+                const tmp_lock = switch (tmp_mcv) {
+                    .register => |reg| self.register_manager.lockRegAssumeUnused(reg),
+                    else => null,
+                };
+                defer if (tmp_lock) |lock| self.register_manager.unlockReg(lock);
+
+                // check if overflow happens
+                try self.genBinOpMir(.{ ._, .cmp }, lhs_ty, tmp_mcv, lhs_mcv);
+                const overflow_reloc = try self.genCondBrMir(lhs_ty, .{ .eflags = Condition.ne });
+
+                // if overflow,
+                // for unsigned integers, the saturating result is just its max
+                // for signed integers,
+                //   if lhs is positive, the result is its max
+                //   if lhs is negative, it is min
+                switch (lhs_ty.intInfo(zcu).signedness) {
+                    .unsigned => {
+                        const bound_mcv = try self.genTypedValue(try lhs_ty.maxIntScalar(self.pt, lhs_ty));
+                        try self.genCopy(lhs_ty, dst_mcv, bound_mcv, CopyOptions{});
+                    },
+                    .signed => {
+                        // check the sign of lhs
+                        try self.genBinOpMir(.{ ._, .cmp }, lhs_ty, lhs_mcv, try self.genTypedValue(try self.pt.intValue(lhs_ty, 0)));
+                        const sign_reloc_condbr = try self.genCondBrMir(lhs_ty, .{ .eflags = Condition.g });
```

Comment:
The ZF is not reliable after `genBinOpMir(.{ ._, .cmp }, ..)`, please use a condition that does not depend on it instead.

Reply:
I am going to use one more SHR to get the sign bit and pass it directly as condition. Thank you for the catch (:

It seems that airCmp is also using genBinOpMir with cmp and returning eflags conditions. is this intended?
https://github.com/ziglang/zig/blob/5cfcb015033864c769235726975ab9919c217ef9/src/arch/x86_64/CodeGen.zig#L20662-L20663

Reply:
As you can see [here](https://github.com/ziglang/zig/blob/d5db02728ce3ff3ba74a3e3f9050c3ea9ed34752/src/arch/x86_64/CodeGen.zig#L20444-L20452), all condition depending on ZF are avoided.

Reply:
`eq` and `neq` are also ZF conditions, am I understanding wrong? If they are okay to use, then the current code should be working (the `gt` has been replaced by `sign bit == 1`), or else the equal condition above this piece of code also needs to be replaced.

Reply:
Indeed, `airShlWithOverflow` is also wrong.

Reply:
How about hacking `genBinOpMir` to emit JCC after each `asmRegisterRegister`?

Reply:
If only it were that simple... I'm sure there exists a jcc that would work for each desired condition, but we don't know the desired condition there.

Reply:
Updated to `Temp.cmpInts`. Thank you!